libqrexec/exec.c

```diff
@@ -343,53 +361,61 @@ struct qrexec_parsed_command *parse_qubes_rpc_command(
             goto err;
         }
 
-        if (end - start > MAX_SERVICE_NAME_LEN) {
-            LOG(ERROR, "Command too long (length %zu)",
-                end - start);
+        if (end <= start) {
+            LOG(ERROR, "Service descriptor is empty (too many spaces after QUBESRPC?)");
             goto err;
         }
 
-        cmd->service_descriptor = strndup(start, end - start);
-        if (!cmd->service_descriptor) {
-            PERROR("strndup");
+        size_t const descriptor_len = (size_t)(end - start);
+        if (descriptor_len > MAX_SERVICE_NAME_LEN) {
+            LOG(ERROR, "Command too long (length %zu)", descriptor_len);
             goto err;
         }
 
         /* Parse service name ("qubes.Service") */
 
-        const char *plus = memchr(start, '+', end - start);
-        if (plus) {
-            if (plus - start == 0) {
-                LOG(ERROR, "Service name empty");
-                goto err;
-            }
-            if (plus - start > NAME_MAX) {
-                LOG(ERROR, "Service name too long to execute (length %zu)",
-                    plus - start);
-                goto err;
-            }
-            cmd->service_name = strndup(start, plus - start);
-            if (!cmd->service_name) {
-                PERROR("strndup");
-                goto err;
-            }
-        } else {
-            cmd->service_name = strndup(start, end - start);
-            if (!cmd->service_name) {
-                PERROR("strdup");
-                goto err;
-            }
+        const char *const plus = memchr(start, '+', descriptor_len);
+        size_t const name_len = plus != NULL ? (size_t)(plus - start) : descriptor_len;
+        if (name_len > NAME_MAX) {
+            LOG(ERROR, "Service name too long to execute (length %zu)", name_len);
+            goto err;
+        }
+        if (name_len < 1) {
+            LOG(ERROR, "Service name empty");
+            goto err;
+        }
+        cmd->service_name = malloc(name_len + 1);
+        if (!cmd->service_name) {
+            PERROR("malloc");
+            goto err;
+        }
+        memcpy(cmd->service_name, start, name_len);
+        cmd->service_name[name_len] = '\0';
```

Comment:
Why open-code strndup?

Reply:
`strndup()` has an implicit `strnlen()` (to get the buffer size) which isn’t needed here (dom0 promises there are no intermediate NUL bytes).  I added a `memdupnul()` function that also has a `PERROR` on failure.